src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs

```diff
@@ -304,12 +309,17 @@ protected virtual async Task OnConnectionComplete(IServiceConnection serviceConn
             serviceConnection.ConnectionStatusChanged -= OnConnectionStatusChanged;
 
             var index = ServiceConnections.IndexOf(serviceConnection);
+
             if (index != -1)
             {
-                // first FixedConnectionCount connections are "fixed" and always try to restart
+                // always try to restart first FixedConnectionCount connections
                 if (index < FixedConnectionCount)
                 {
-                    await RestartFixedServiceConnectionCoreAsync(index);
+                    // unless the container is offline or stopped
+                    if (ReadyForNewConnections)
```

Comment:
too much `if`, how about `if (index < FixedConnectionCount && !_terminated)`

Reply:
Yeah, initially I did exactly what you proposed. This (in case of `_terminated` == true ) will result in calling `RemoveOnDemandConnection` in else statement and eventually emptying ServiceConnections. This is ok, but I was worrying about our usage of [Task.WhenAny on ServiceConnections](https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionContainerBase.cs#L222) which will throw when the collection is finally empty. Even though right now I don't see a scenario when someone would check `ConnectionInitializedTask` during shutdown stage I thought I will save the potential trouble and keep the old behavior of not removing stopped connections from `ServiceConnections`.